dace/sourcemap.py

```diff
@@ -0,0 +1,344 @@
+        path = os.path.abspath(
+            folder +
+            "/map/map_" + language + ".json"
+        )
+
+        with open(path, "w") as json_file:
+            json.dump(self.map, json_file, indent=4)
+
+        return os.path.abspath(folder)
+
+
+class MapCpp(MapCreater):
+    def __init__(self, code, name, target_name):
+        super(MapCpp, self).__init__(name)
+        self.code = code
+        self.mapper()
+        self.map['target'] = target_name
+
+        folder = self.save("cpp")
+
+        tmp = get_tmp(self.name)
+        remove_tmp(self.name)
+
+        # Send information about the function to VSCode
+        if tmp is not None:
+            send({
+                    "type": "registerFunction",
+                    "name": self.name,
+                    "path_cache": folder,
+                    "path_file": tmp.get("src_file"),
+                    "target_name": target_name
+                })
+
+    def mapper(self):
+        """ For each line of code retrieve the corresponding nodes
+            and map the nodes to this line
+        """
+        for line_num, line in enumerate(self.code.split("\n"), 1):
+            nodes = self.get_node(line)
+            for node in nodes:
+                self.create_mapping(node, line_num)
+
+    def create_mapping(self, node, line_num):
+        if node.sdfg_id not in self.map:
+            self.map[node.sdfg_id] = {}
+        if node.state_id not in self.map[node.sdfg_id]:
+            self.map[node.sdfg_id][node.state_id] = {}
+
+        state = self.map[node.sdfg_id][node.state_id]
+        for node_id in node.node_ids:
+            if node_id not in state:
+                state[node_id] = {
+                    'from': line_num,
+                    'to': line_num
+                }
+            elif state[node_id]['to'] + 1 == line_num:
+                # If the current node maps to the previous line 
+                # (before "line_num"), then extend the range this node maps to
+                state[node_id]['to'] += 1
+
+    def get_node(self, line):
+        """ Retrive all identifiers set at the end of the line of code.
+            Example: x = y ////__DACE:0:0:0 ////__DACE:0:0:1
+                Returns [SDFGL(0,0,0), SDFGL(0,0,1)]
+        """
+        line_identifiers = self.get_identifiers(line)
+        nodes = []
+        for identifier in line_identifiers:
+            ids_split = identifier.split(":")
+            nodes.append(
+                SdfgLocation(
+                    ids_split[1],
+                    ids_split[2],
+                    # node might be an edge
+                    ids_split[3].split(",")
+                )
+            )
+        return nodes
+
+    def get_identifiers(self, line):
+        line_identifier = re.findall(
+            r'(\/\/\/\/__DACE:[0-9]:[0-9]:[0-9](,[0-9])*)',
+            line
+        )
+        # remove tuples
+        return [x or y for (x, y) in line_identifier]
```

Comment:
strange behavior. Could you document with an example?

Reply:
Currently, the regex expression returns a list of 2 groups (list of tuples) where one element is empty. This is due to the identifiers also indicating edges ////DACE__0:0:0,2. I'm considering either ignoring edges, just taking the from node of the edge or I'll try to rewrite the regex expression to only return one group (I'm not very skilled in regex expressions)

Reply:
my comment would be just to add a comment above the list comprehension saying "This ignores edges of the format ..." or so